### ISSUE TITLE
overriding standard types

### ISSUE THREAD

**Issue by nadako** [HaxeFoundation/hxnodejs#42]
I tried to add `haxe.io.Bytes` implementation, but there's a problem that in a macro it also tries to use the overriden class and then complains about using `js.node.Buffer` in a macro context.

What can we do with that? One option is to just move nodejs sys stuff into Haxe itself (and fence with `#if nodejs`), another one is to make it possible to override types only for target output context, not the macro one.

Also, `@:coreApi` metadata doesn't work on our overriden types.

cc @Simn @ncannasse

**Comment by nadako** [comment]
Bytes implementation https://github.com/HaxeFoundation/hxnodejs/commit/72d7b831dba73317f2b0e725f975ad83f8bea324

**Comment by nadako** [comment]
Oh, and thanks to @andyli and Travis CI we can see that `@:coreApi` actually worked in 3.2.0: https://travis-ci.org/HaxeFoundation/hxnodejs/jobs/79463579

**Comment by eduardo-costa** [comment]
I didn't get yet whay @coreApi does :)
Em 09/09/2015 10:05, "Dan Korostelev" <notifications@github.com> escreveu:

**Comment by nadako** [comment]
@eduardo-costa it checks whether all fields are defined correctly in the overriden std class. it's used in haxe std for providing target implementations. it does check field existance, type, access modifiers, method signatures, etc.

**Comment by eduardo-costa** [comment]
Got it! Thanks!
We should describe more of thia in he docs!
Em 09/09/2015 11:23, "Dan Korostelev" <notifications@github.com> escreveu:

**Comment by ncannasse** [comment]
Yes I'm aware of the issue with Bytes redefinition. I actually opened https://github.com/HaxeFoundation/haxe/issues/4249 because of that. Not sure if that's the right solution or not.

I actually thing we need to find better ways for libraries to override/implement some std classes as well as configure which implementation is used depending on the compiled platform (in a safe manner, ie like we do for `@:coreApi`in stdlib)

**Comment by waneck** [comment]
I wish we implemented the `_std` magic for all classpaths - not just the std classpath

**Comment by ncannasse** [comment]
This currently breaks haxe.io.BytesBuffer compilation, and breaks some WebGL code I'm having in Node Webkit as well (since I need to have Bytes that maps to JS ArrayBuffer). Should we not drop the haxe.io.Bytes <-> NodeBuffer support since it seems recent NodeJS version does uses ArrayBuffer ? (I didn't test the later)

**Comment by ncannasse** [comment]
Up again : should we remove this and instead add NodeBuffer <-> Bytes conversion (in NodeBuffer class) since they are both based on ArrayBuffer but uses different APIs ?

**Comment by nadako** [comment]
To be honest, I'd remove anything but node.js externs (to reduce scope of this particular lib), and then make a separate lib that provides sys stuff.

**Comment by clemos** [comment]
@nadako +1

**Comment by ncannasse** [comment]
Maybe, but that just moves the problem away, no ?
I think it would be nice for Haxe users to have a crossplatform sys API for Node. Even more if that's a complex problem to solve so that users don't have to reinvent the wheel.

But looking at the Buffer API it seems there's no way to easy convert from/to ArrayBuffer (with no copying being done in the process). I don't then understand why they say that in 4.0+, Buffer is backed by UInt8Array. Unless there's a trick, can Node users confirm this ?

If there's no easy way, I agree we should move it out of hxnodejs for now, until we come up with a solution.

**Comment by nadako** [comment]
Looking at this code: https://github.com/nodejs/node/blob/master/lib/buffer.js#L43-L62, it seems that Buffer actually IS Uint8Array

**Comment by ncannasse** [comment]
Oh. then take makes things easy, right ?

**Comment by nadako** [comment]
I don't know, will have to look into it? What's the decision anyway? Support 4.0+ and provide only `Sys`/`sys.*` stuff?

**Comment by ncannasse** [comment]
What I'm proposing (open for discuss):
- support 4.0+ Buffer (which extends UInt8Array)
- do not redefine haxe.io.Bytes and especially not BytesData (breaks several std classes and 3rd party libraries)
- add API in Buffer (using inline methods?) which allows from/to Bytes, I can help with this
- implement sys/Sys

**Comment by nadako** [comment]
Sounds sane, I'll try to find time to look into it today.

**Comment by ncannasse** [comment]
BTW it might be possible to have the node Buffer be a reference to an ArrayBuffer (with memory sharing) if created by passing an ArrayBuffer, to confirm by experiment, see
https://github.com/nodejs/node/blob/master/lib/buffer.js#L142

**Comment by ncannasse** [comment]
(passing an UInt8Array for instance will create a copy)

**Comment by nadako** [comment]
Okay, I removed haxe.io stuff and added conversion helper methods. Please review.

**Comment by nadako** [comment]
We have a problem with Haxe 3.2.0 however: the `byteLength` static method of node's `Buffer` class has the same name as `Uint8Array` non-static property, so it fails to compile (https://travis-ci.org/HaxeFoundation/hxnodejs/jobs/88176317). It's allowed on Haxe development though.

**Comment by nadako** [comment]
Also, can we get rid of js.html.compat stuff for nodejs somehow?

**Comment by eduardo-costa** [comment]
@nadako +1 to keep hxnodejs extern-only
We have the `hxnodelibs` repo that can fit extra libs for `sys` and externs for other npm tools.

**Comment by clemos** [comment]
@nadako meanwhile you can : 
```
inline static function _byteLength ( str : String, ?enc : String ) : Int return untyped Buffer['byteLength'](str,enc);
```

**Comment by nadako** [comment]
Yeah, we probably should do that for now and then deprecate it when 3.3.0 comes around

**Comment by nadako** [comment]
I did that, pls review!

**Comment by clemos** [comment]
Looks good to me

**Comment by ncannasse** [comment]
ok, everything seems good to me!
shall we move forward with a proper release?
or do anyone want to contribute more sys/Sys implementation (quickly) before that?

Also, we need to decide on the name. We can have the library named either "nodejs" or "hxnodejs", as soon as we have a proper `-D nodejs` added as well so users can `#if nodejs ...`. In all cases, it will be the best to replace / take down no longer maintained node libraries at the same time, such as  http://lib.haxe.org/p/nodejs and http://lib.haxe.org/p/nodehx  if their respective authors agree with it?

**Comment by eduardo-costa** [comment]
I'm good!
I own the nodehx lib! Feel free to put it down!


2015-11-03 18:18 GMT-02:00 Nicolas Cannasse <notifications@github.com>:

> ok, everything seems good to me!
> shall we move forward with a proper release?
> or do anyone want to contribute more sys/Sys implementation (quickly)
> before that?
>
> Also, we need to decide on the name. We can have the library named either
> "nodejs" or "hxnodejs", as soon as we have a proper -D nodejs added as
> well so users can #if nodejs .... In all cases, it will be the best to
> replace / take down no longer maintained node libraries at the same time,
> such as http://lib.haxe.org/p/nodejs and http://lib.haxe.org/p/nodehx if
> their respective authors agree with it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/HaxeFoundation/hxnodejs/issues/42#issuecomment-153475059>
> .
>



-- 
[image: Imagem inline 1]

**Comment by nadako** [comment]
@ncannasse 
> Also, can we get rid of js.html.compat stuff for nodejs somehow?

Because any Buffer usage right now outputs a ton of unused compat stuff in the output js.

**Comment by ncannasse** [comment]
@nadako Sorry, I've let this pass. Can you not simply #if !nodejs it ?

**Comment by dionjwa** [comment]
You can have http://lib.haxe.org/p/nodejs!

**Comment by nadako** [comment]
Okay, I added `#if !nodejs` guards to haxe std, seems quite clean (although it still adds empty `haxe.io.Bytes` constructor for some reason, but we can live with it, I guess). As for the name, I'd release it as `hxnodejs` and upload a deprecation release for nodejs and nodehx that point to the new lib.

**Comment by kevinresol** [comment]
hxnodejs seems consistent with hxjava, hxcs, hxcpp

**Comment by nadako** [comment]
Honestly, I'm okay with any name.

**Comment by kevinresol** [comment]
what if we name it hxnodejs and confine its scope as node extern + haxe std library?
Anything else will go to another nodelib repo

**Comment by eduardo-costa** [comment]
We have hxnodelibs for that..
I vote for hxnodejs as extern-only
Em 04/11/2015 23:38, "Kevin Leung" <notifications@github.com> escreveu:

**Comment by nadako** [comment]
@kevinresol +1

If there'll be no further discussion, I can make a haxelib release.

**Comment by andyli** [comment]
@nadako One question: Does the JS output runs equally fine on browsers when `-D nodejs`? Ref. https://github.com/HaxeFoundation/haxe/issues/4524

**Comment by nadako** [comment]
Oh, I gotta read that...

**Comment by nadako** [comment]
So, that topic seems to be Haxe-related which means it shouldn't block release of hxnodejs. @ncannasse waiting for you to say the word.

**Comment by nadako** [comment]
I'll close this issue, as we really should use #40 for release discussion. The `Bytes` problem is solved now.

**Comment by nadako** [comment]
Actually, let's leave this open for now, since the general issue is still there. Maybe it should go to Haxe's issues?

**Comment by ncannasse** [comment]
hxnodejs is ok for me. Let's release! (unless there's something else to discuss before that?)

**Comment by back2dos** [comment]
@nadako The issue with the type collision can be solved by making two implementations in `js.node.Bytes` and `js.node.BytesData` and then call some `--macro` in `extraParams.hxml` to  `Context.defineType` to alias `haxe.io.Bytes` and `haxe.io.BytesData` to them. That way the implementation will not leak into macro context.
